cli/loader_lib.c

```diff
@@ -153,6 +152,206 @@ JL_DLLEXPORT const char * jl_get_libdir()
+    size_t allocated = newlen + reserve + 1;
+    char *newname = (char *)malloc(allocated);
+    if (!newname) {
+        fprintf(stderr, "Out of memory.\n");
+        exit(1);
+    }
+    strcpy(newname, libdir);
+    strcpy(newname + ldlen, append);
+    newname[allocated] = '\0';
+    *ret = newname;
+    *len = newlen;
+}
+
+#ifdef _OS_LINUX_
+#ifndef GLIBCXX_LEAST_VERSION /* Should always be defined in the makefile. This appeases the linter. */
+#define GLIBCXX_LEAST_VERSION "GLIBCXX_a.b.c"
+#endif
+
+#include <link.h>
+#include <sys/wait.h>
+
+static void write_wrapper(int fd, char *str, size_t len)
+{
+    size_t written_sofar = 0;
+    while (len) {
+        ssize_t bytes_written = write(fd, str + written_sofar, len);
+        if (bytes_written == -1 && errno == EINTR) continue;
+        if (bytes_written == -1 && errno != EINTR) {
+            fputs("Error during libstdcxxprobe in child process:", stderr);
+            perror("write");
+            fflush(stderr);
+            _exit(1);
+        }
+        len -= bytes_written;
+        written_sofar += bytes_written;
+    }
+}
+
+// Where the buf passed in was malloced
+static void read_wrapper(int fd, char **ret, size_t *ret_len)
+{
+    // Allocate an initial buffer
+    size_t len = JL_PATH_MAX;
+    char *buf = (char *)malloc(len + 1);
+    if (!buf) {
+        fprintf(stderr, "Error during libstdcxxprobe in parent process:\n");
+        fprintf(stderr, "malloc() returned NULL.\n");
+        exit(1);
+    }
+
+    // Read into it, reallocating as necessary
+    size_t have_read = 0;
+    while (1) {
+        ssize_t n = read(fd, buf + have_read, len - have_read);
+        have_read += n;
+        if (n == 0) break;
+        if (n == -1 && errno != EINTR) {
+            fputs("Error during libstdcxxprobe in parent process:", stderr);
+            perror("read");
+            exit(1);
+        }
+        if (n == -1 && errno == EINTR) continue;
+        if (have_read == len) {
+            buf = (char *)realloc(buf, 1 + (len *= 2));
+            if (!buf) {
+                fputs("Error during libstdcxxprobe in parent process:", stderr);
+                fprintf(stderr, "malloc() returned NULL.\n");
+                exit(1);
+            }
+        }
+    }
+
+    *ret = buf;
+    *ret_len = have_read;
+}
+
+// On Linux, it can happen that the system has a newer libstdc++ than the one we ship,
+// which can break loading of some system libraries: <https://github.com/JuliaLang/julia/issues/34276>.
+
+// Return the path to the libstdcxx to load. Otherwise, print the error and exit.
+// The path returned must be freed.
+static char *libstdcxxprobe(void)
+{
+    // Create the pipe and child process.
+    int fork_pipe[2];
+    int ret = pipe(fork_pipe);
+    if (ret == -1) {
+        perror("Error during libstdcxxprobe:\npipe");
+        exit(1);
+    }
+    // Make sure the buffers are clean before forking.
+    fflush(stdout);
+    fflush(stderr);
```

Comment:
Some of the error handling does use `fprintf()`, and my idea was that:
1. There's only one thread right now, so there's no way to partially flush or confuse the order of writes to the stdio buffers
2. I want the stdio buffers to be empty so that after the `fork()` its contents won't be duplicated

Is the first point not necessarily the case? If it's not, how about calling `flockfile()`?

Reply:
It looked like you were just dumping strings, so write would suffice

Reply:
Now that all the error handling in the parent process uses `write()`, what should be done with the stuff that's in the meantime to the std streams? Should the parent and child both call `exit()`, or `_exit()`?

Reply:
child always uses `_exit`, parent always uses `exit`